### PR TITLE
Created MainScraper

### PR DESCRIPTION
**Pull request by btkcodedev**
### Backend Scraper for flipkart
_Libraries of Beautifulsoup and requests are used for scraping the contents_ 

- The scraper make use of the Flipkart site in which 2 types of styles used in the sections are scraped

- As an example, Laptops, mobiles, Cameras that uses the style 1 and headphones, Air conditioners uses the style 2 are scraped,
Both the styles can scrap several other categories.

> Logic
> The styles in the flipkart mainly divided into two and both are scraped on the basis of the tag name and pattern
> If the first style doesn't match the pattern, then no result will be forwarded to the list, thus the second styling is checked

- The number of pages to scrap can be changed to the limit in line 10
`for i in range(1,5):  #use n+1 pages for clearing index out of range error`

- The number of pages can be set in the line 24 as for loop range
`for i in range(1,limit):  #only use n-1 pages which are proxy safe and error free in the list`
  `myurl= urlss[i]`

- The results will be the Link,Title,Rating,Amount that will be listed in different python lists in strict order along with number of things scraped at last. (Also the csv and io are imported which can be used when file management is necessary)
`print(l) #Links`
`print(title) #Title`
`print(rat) #Rating`
`print(amount) #Amount`
`print(len(amount))  #Number of items scraped`
`print("pages")`

- The frontend can make use of the variables in the URL in line 16 for changing the need for the user along some dropdown boxes 
`urlss=urlss+[cams]  # Change the variable with a dropdown box when merging with its front end`

-Hope it gives an addition for the project.

### CODE REVIEW

**Comment by anandbaburajan** [comment]
@btkcodedev Hey! Thanks for the contribution! The comments helped a lot. I've a few suggestions, since the project maintainer and other members don't seem to be active:

You can refactor `scaper.py` so that it can be used as the back-end of this project . You can do that by modularizing the existing code into their own functions. Example: `get_urls()`, `scrape_page()` etc. and replacing the `print` statements with `return`s. The functions can take parameters like number of pages, product category, etc. This way, your functions can be easily used by the front-end devs.  :) You can also rename the file to `flipkart_scraper.py` right? Also, it's better if you put the comments above the statements, instead of in-line. Also, no need for comments for the `import` statements.

Also, next time onwards, please make sure to create a new branch for your PRs.

Thanks again!

**Comment by btkcodedev** [comment]
@AnandBaburajan 
The changes are made to the code
Also it can be interpreted and results can be examined by the front end developers

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: Put comment on top of the statement here and similarly for all other statements.

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: Creating a dictionary which stores all those links and then accessing the values would look better. Also, since `for i in range(1,limit)` is repeated a lot of times here. Maybe you can refactor the code so that it's used just once. (You can put the conditions inside the loop). Also, you can use `switch - case` here with a `default`. Would be more readable.

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: It'd be better if the Python variables were in camel_case. Like, `dropdown_selection`

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: A good practice is top have all of your import statements at the top of your file, instead of inside functions. You can also remove the comments because interested people can search for that on their own. (better looking)

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: The best Pythonic way of checking empty list: `if not title:`

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: Since these are lists, please give more meaningful names likes `titles`, `amounts` and please give better names to the rest

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: The `-1` looks a bit weird, can you refactor the code to deal with this?

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: Do we need these `prints`?

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: Please use a formatter for Python to take care of consisten spacing between lines and variables, etc.

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: Also, it's `get_urls` right?

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: Also, can you look into refactoring the part hereafter into its own function? Would help in reducing the code complexities (you can check your PRs codeclimate results)

**Review by anandbaburajan — CHANGES_REQUESTED** [review]
Thanks! Great work~ Looking forward to see this integrated with the client side!

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: Also can you add a docstring which helps the user to understand what this function takes and returns? Thanks!

**Comment by btkcodedev** [comment]
@AnandBaburajan The required changes are done
Now the whole code can be called with one function
Eg; scrape_page(get_urls(4,6),6)
Thus it will be easier for the front end to merge
More details are shared in the latest commit

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py:131]: Did you look into making this into its own function? Would help with decreasing the complexity (check the codeclimate results). I think `scrape_page` can be split into two-three function! Thanks!

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: Like I mentioned earlier, can you please add a docstring for the functions? https://www.geeksforgeeks.org/python-docstrings/

**Review by anandbaburajan — CHANGES_REQUESTED** [review]
@btkcodedev Thanks! Rest of the changes look good! I've added a few comments then this PR can be merged!

**Comment by btkcodedev** [comment]
@AnandBaburajan The required changes are done,
- Complexity reduced:
The scrape_page() was split into two and complexity reduced from 46 to 14 and 24 for two functions.
But that will be not a problem, I think because the second type function will only execute if the first type fails
And it totally depends on the number of pages scraped, and that will be not more than 10 I guess.
- Docstrings added:
Also now the frontend developers can easily identify the flow using the docstrings and comments

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: The best way to use doctrings is the following format:

```python
def my_function(arg1): 
    """ 
    Summary line. 
  
    Extended description of function. 
  
    Parameters: 
    arg1 (int): Description of arg1 
  
    Returns: 
    int: Description of return value 
  
    """
  
    return arg1 
```

Also, "unlike conventional source code comments, the docstring should describe what the function does, not how". So here, it would be better if you describe in simple terms what `get_urls` takes as input and what it returns. Rest of the "good to know" things can be put in a comment somewhere else.

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: Similarly fix the docstring here.

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: Can you change it to `scrape_page_type_2`? Would be more readable I think.

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: Can you change it to `scrape_page_type_1`? Would be more readable I think.

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: Similarly fix the docstring here.

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: This can be a simple comment because docstrings are only used to describe functions, classes, etc :)

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: Can you put it somwehere at the top?

**Review by anandbaburajan — CHANGES_REQUESTED** [review]
@btkcodedev I've added some final minor comments after which this would be good to merge! Thanks!

**Comment by btkcodedev** [comment]
@AnandBaburajan Required changes are done

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py:17]: Please put the `"""` on the top and below the description. Also, I see you provided a docstring for the return statement below which is incorrect. I suggest to go with something like:

```python
def get_urls(DropDownSelection, limit):
    """ 
    Function to make a list of links according to the limit and selection.
  
    Parameters: 
    DropDownSelection (int): Selects the category from the frontend
    limit (int): The limit for the number of web pages selected by user or by developer in frontend
  
    Returns: 
    urlss (List): List names urlss that contains the list of links based on the given category.
    """
  
    return arg1 
```

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py:40]: Fix docstring here

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py:32]: Docstrings are only used to describe classes, functions etc. Please remove this one.

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: Fix docstring as suggested

File: flipkart_scraper.py
Comment (anandbaburajan) [inline flipkart_scraper.py]: Fix docstring as suggested

**Review by anandbaburajan — CHANGES_REQUESTED** [review]
@btkcodedev I've added some comments. Please go through https://www.geeksforgeeks.org/python-docstrings/ for more info on docstrings.

File: flipkart_scraper.py
Comment (anandbaburajan) [reply flipkart_scraper.py:17]: @btkcodedev please correct the docstring as suggested in the example

File: flipkart_scraper.py
Comment (anandbaburajan) [reply flipkart_scraper.py:32]: Docstrings aren't used here, please remove this and refer to the article I shared earlier

**Pull request closed without merging**
